lib/Dialect/Handshake/HandshakeTypes.cpp

```diff
@@ -99,9 +103,152 @@ checkChannelExtra(function_ref<InFlightDiagnostic()> emitError,
   return success();
 }
 
+/// Parse the extra signals from the IR representation.
+/// Due to the restriction on the joint parser of ControlType and ChannelType,
+/// we don't parse square brackets here.
+/// For example: spec: i1, tag: i32
+/// This function is common to both ControlType and ChannelType.
+static LogicalResult
+parseExtraSignals(function_ref<InFlightDiagnostic()> emitError,
+                  AsmParser &odsParser,
+                  SmallVector<ExtraSignal::Storage> &extraSignalsStorage) {
+
+  auto parseSignal = [&]() -> ParseResult {
+    auto &signal = extraSignalsStorage.emplace_back();
+
+    if (odsParser.parseKeywordOrString(&signal.name) ||
+        odsParser.parseColon() || odsParser.parseType(signal.type))
+      return failure();
+
+    // Attempt to parse the optional upstream symbol
+    if (!odsParser.parseOptionalLParen()) {
+      std::string upstreamSymbol;
+      if (odsParser.parseKeywordOrString(&upstreamSymbol) ||
+          upstreamSymbol != UPSTREAM_SYMBOL || odsParser.parseRParen())
+        return failure();
+      // The default value of `signal.downstream` is true.
+      // I think this is a bit confusing.
+      // TODO: signal.upstream?
```

Comment:
This kind of comment should not be in the code, if you want to raise those questions do so in the PR itself. I also don't see the problem here, all extra signals examples we have currently are downstream signals so I made the default the common use case.

Reply:
I meant to start a conversation about this in the PR later, but I forgot. I removed this comment from the file.
While this doesn't cause any bugs, I prefer the upstream flag, as I thought flag variables are typically set to true in special cases (possibly this notion comes from the fact that bool variables are initialized to false by default).
Here downstream is the general case, as you mentioned.
Overwriting the default value of this flag to true and conditionally setting the value to false later felt a bit convoluted to me.

Reply:
> possibly this notion comes from the fact that bool variables are initialized to false by default

Native types like bool do not have a default init value in C++ unless you give them one explicitly, it's in fact a common source of bugs to leave a variable like this uninitialized. For example in the following the value of `x` is "whatever was on the stack when this function was called".

```cpp
bool getPseudoRandomBool() {
  bool x;
  return x;
}
```

Reply:
Wow, I didn't know this, thanks.
Now I understand that even if the default value of the flag is false it needs to be explicitly initialized.